src/shared/sd-dns-resolver.c

```diff
@@ -0,0 +1,354 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "dns-resolver-internal.h"
+
+#include "macro.h"
+#include "unaligned.h"
+#include "socket-netlink.h"
+#include "string-table.h"
+#include "string-util.h"
+#include "strv.h"
+
+void sd_dns_resolver_done(sd_dns_resolver *res) {
+        res->auth_name = mfree(res->auth_name);
+        res->addrs = mfree(res->addrs);
+        res->dohpath = mfree(res->dohpath);
+}
+
+sd_dns_resolver *sd_dns_resolver_unref(sd_dns_resolver *res) {
+        sd_dns_resolver_done(res);
+        return mfree(res);
+}
+
+void sd_dns_resolver_array_free(sd_dns_resolver resolvers[], size_t n) {
+        assert(resolvers || n == 0);
+
+        FOREACH_ARRAY(res, resolvers, n)
+                sd_dns_resolver_done(res);
+
+        free(resolvers);
+}
```

Comment:
i am a bit confused by this. this is a public API (as per the sd_ prefix), but it takes an array of structs (as opposed to an array of pointers to structs). This means the sd_dns_resolver_resolvers[] structure would have to be open coded for this to be callable from others. But we really don#t want open coded structs usually...

Reply:
Err, yeah this function doesn't really need to be public. I'll just move it to the private header.